File: 13/app/src/main/java/com/example/studentcoursemanager/MainActivity.java
package com.example.studentcoursemanager;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {
    private CourseAdapter adapter;
    private SearchView searchView;
    private String currentQuery = "";

    private final Runnable courseListener = this::renderCourses;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setSupportActionBar(findViewById(R.id.topAppBar));

        adapter = new CourseAdapter(
                this::openDetail,
                this::openEdit,
                this::confirmDelete
        );

        RecyclerView recyclerView = findViewById(R.id.recyclerCourses);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);

        FloatingActionButton fabAddCourse = findViewById(R.id.fabAddCourse);
        fabAddCourse.setOnClickListener(view -> openAddCourse());

        CourseRepository.addListener(courseListener);
        renderCourses();
    }

    @Override
    protected void onDestroy() {
        CourseRepository.removeListener(courseListener);
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        MenuItem searchItem = menu.findItem(R.id.action_search);
        searchView = (SearchView) searchItem.getActionView();
        searchView.setQueryHint("Search by name or code");
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                currentQuery = query == null ? "" : query;
                renderCourses();
                searchView.clearFocus();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                currentQuery = newText == null ? "" : newText;
                renderCourses();
                return true;
            }
        });
        searchView.setOnCloseListener(() -> {
            currentQuery = "";
            renderCourses();
            return false;
        });
        return true;
    }

    private void renderCourses() {
        List<Course> allCourses = CourseRepository.getCourses(this);
        List<Course> filteredCourses;
        if (currentQuery.trim().isEmpty()) {
            filteredCourses = allCourses;
        } else {
            String query = currentQuery.trim().toLowerCase(Locale.ROOT);
            filteredCourses = allCourses.stream()
                    .filter(course -> course.getName().toLowerCase(Locale.ROOT).contains(query)
                            || course.getCode().toLowerCase(Locale.ROOT).contains(query))
                    .toList();
        }

        adapter.submitList(filteredCourses);

        RecyclerView recyclerView = findViewById(R.id.recyclerCourses);
        ConstraintLayout emptyStateView = findViewById(R.id.emptyStateView);
        TextView emptyTitle = findViewById(R.id.textEmptyTitle);

        boolean hasCourses = !allCourses.isEmpty();
        boolean hasFilterResult = !filteredCourses.isEmpty();

        recyclerView.setVisibility(hasCourses && hasFilterResult ? android.view.View.VISIBLE : android.view.View.GONE);
        emptyStateView.setVisibility(!hasCourses || !hasFilterResult ? android.view.View.VISIBLE : android.view.View.GONE);
        emptyTitle.setText(hasCourses ? "No matching courses found" : "No courses added yet");
    }

    private void openAddCourse() {
        startActivity(new Intent(this, AddCourseActivity.class));
    }

    private void openEdit(Course course) {
        startActivity(EditCourseActivity.newIntent(this, course));
    }

    private void openDetail(Course course) {
        startActivity(CourseDetailActivity.newIntent(this, course));
    }

    private void confirmDelete(Course course) {
        new MaterialAlertDialogBuilder(this)
                .setTitle("Delete course")
                .setMessage("Delete " + course.getName() + " from your list?")
                .setNegativeButton("Cancel", null)
                .setPositiveButton("Delete", (dialog, which) -> CourseRepository.deleteCourse(this, course.getId()))
                .show();
    }
}

File: 5/app/src/main/java/com/example/contactbookapp/MainActivity.kt
package com.example.contactbookapp

import android.app.AlertDialog
import android.os.Bundle
import android.widget.*
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var adapter: ContactAdapter
    private val contactList = mutableListOf<Contact>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val listView = findViewById<ListView>(R.id.listView)
        val searchView = findViewById<SearchView>(R.id.searchView)
        val fab = findViewById<com.google.android.material.floatingactionbutton.FloatingActionButton>(R.id.fab)
        val emptyView = findViewById<TextView>(R.id.emptyView)

        adapter = ContactAdapter(this, contactList)
        listView.adapter = adapter
        listView.emptyView = emptyView

        // FAB Add Contact
        fab.setOnClickListener {
            showAddContactDialog()
        }

        // Item Click
        listView.setOnItemClickListener { _, _, position, _ ->
            val c = contactList[position]
            Toast.makeText(this,
                "Name: ${c.name}\nPhone: ${c.phone}\nEmail: ${c.email}",
                Toast.LENGTH_LONG).show()
        }

        // Long Press Delete
        listView.setOnItemLongClickListener { _, _, position, _ ->
            AlertDialog.Builder(this)
                .setTitle("Delete Contact")
                .setMessage("Are you sure?")
                .setPositiveButton("Yes") { _, _ ->
                    contactList.removeAt(position)
                    adapter.notifyDataSetChanged()
                }
                .setNegativeButton("No", null)
                .show()
            true
        }

        // Search Filter
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?) = false

            override fun onQueryTextChange(newText: String?): Boolean {
                adapter.filter(newText ?: "")
                return true
            }
        })
    }

    private fun showAddContactDialog() {
        val view = layoutInflater.inflate(android.R.layout.simple_list_item_2, null)

        val nameInput = EditText(this)
        nameInput.hint = "Name"

        val phoneInput = EditText(this)
        phoneInput.hint = "Phone"

        val emailInput = EditText(this)
        emailInput.hint = "Email"

        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.VERTICAL
        layout.setPadding(40, 20, 40, 10)
        layout.addView(nameInput)
        layout.addView(phoneInput)
        layout.addView(emailInput)

        AlertDialog.Builder(this)
            .setTitle("Add Contact")
            .setView(layout)
            .setPositiveButton("Add") { _, _ ->
                val name = nameInput.text.toString()
                val phone = phoneInput.text.toString()
                val email = emailInput.text.toString()

                if (name.isNotEmpty() && phone.isNotEmpty()) {
                    contactList.add(Contact(name, phone, email))
                    adapter.notifyDataSetChanged()
                }
            }
            .setNegativeButton("Cancel", null)
            .show()
    }
}